protocol/messenger_profile_showcase.go

```diff
@@ -86,3 +124,193 @@ func (m *Messenger) GetProfileShowcasePreferences() (*ProfileShowcasePreferences
+		err = proto.Unmarshal(entriesData, entries)
+		if err != nil {
+			return nil, err
+		}
+
+		return entries, nil
+	}
+
+	// Return empty if no matching key found
+	return &protobuf.ProfileShowcaseEntries{}, nil
+}
+
+func (m *Messenger) GetProfileShowcaseForSelfIdentity() (*protobuf.ProfileShowcase, error) {
+	preferences, err := m.GetProfileShowcasePreferences()
+	if err != nil {
+		return nil, err
+	}
+
+	forEveryone := &protobuf.ProfileShowcaseEntries{
+		Communities:  toProfileShowcaseUpdateEntries(preferences.Communities, ProfileShowcaseVisibilityEveryone),
+		Accounts:     toProfileShowcaseUpdateEntries(preferences.Accounts, ProfileShowcaseVisibilityEveryone),
+		Collectibles: toProfileShowcaseUpdateEntries(preferences.Collectibles, ProfileShowcaseVisibilityEveryone),
+		Assets:       toProfileShowcaseUpdateEntries(preferences.Assets, ProfileShowcaseVisibilityEveryone),
+	}
+
+	forContacts := &protobuf.ProfileShowcaseEntries{
+		Communities:  toProfileShowcaseUpdateEntries(preferences.Communities, ProfileShowcaseVisibilityContacts),
+		Accounts:     toProfileShowcaseUpdateEntries(preferences.Accounts, ProfileShowcaseVisibilityContacts),
+		Collectibles: toProfileShowcaseUpdateEntries(preferences.Collectibles, ProfileShowcaseVisibilityContacts),
+		Assets:       toProfileShowcaseUpdateEntries(preferences.Assets, ProfileShowcaseVisibilityContacts),
+	}
+
+	forIDVerifiedContacts := &protobuf.ProfileShowcaseEntries{
+		Communities:  toProfileShowcaseUpdateEntries(preferences.Communities, ProfileShowcaseVisibilityIDVerifiedContacts),
+		Accounts:     toProfileShowcaseUpdateEntries(preferences.Accounts, ProfileShowcaseVisibilityIDVerifiedContacts),
+		Collectibles: toProfileShowcaseUpdateEntries(preferences.Collectibles, ProfileShowcaseVisibilityIDVerifiedContacts),
+		Assets:       toProfileShowcaseUpdateEntries(preferences.Assets, ProfileShowcaseVisibilityIDVerifiedContacts),
+	}
+
+	mutualContacts := []*Contact{}
+	iDVerifiedContacts := []*Contact{}
+
+	m.allContacts.Range(func(_ string, contact *Contact) (shouldContinue bool) {
+		if contact.mutual() {
+			mutualContacts = append(mutualContacts, contact)
+			if contact.IsVerified() {
+				iDVerifiedContacts = append(iDVerifiedContacts, contact)
+			}
+		}
+		return true
+	})
+
+	forContactsEncrypted, err := m.EncryptProfileShowcaseEntriesWithContactPubKeys(forContacts, mutualContacts)
+	if err != nil {
+		return nil, err
+	}
+
+	forIDVerifiedContactsEncrypted, err := m.EncryptProfileShowcaseEntriesWithContactPubKeys(forIDVerifiedContacts, iDVerifiedContacts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &protobuf.ProfileShowcase{
+		ForEveryone:           forEveryone,
+		ForContacts:           forContactsEncrypted,
+		ForIdVerifiedContacts: forIDVerifiedContactsEncrypted,
+	}, nil
+}
+
+func (m *Messenger) BuildProfileShowcaseFromIdentity(senderPubKey *ecdsa.PublicKey, message *protobuf.ProfileShowcase) (*identity.ProfileShowcase, error) {
```

Comment:
@MishkaRogachev do we protect somehow that messages meant to be delivered to verified contacts can not be revealed by mutual contacts?

Reply:
Yes, each group encrypted with with explicitly specified contact's keys, same as  in identity image flow